compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+bool structContainsOnlyPrimitiveScalars(Type* t);
+bool isSupportedType(Type* t);
+void verifyPrototype(FnSymbol* fn);
+Type* getTypeFromFormal(ArgSymbol* as);
+Type* getTypeFromFormal(FnSymbol* fn, int i);
+bool typeRequiresAllocation(Type* t);
+
+private:
+
+std::string genMarshalBodyPrimitiveScalar(Type* t, bool out);
+std::string genMarshalBodyStringC(Type* t, bool out);
+std::string genComment(const char* msg, const char* pfx="");
+std::string genNote(const char* msg);
+std::string genTodo(const char* msg);
+std::string genHeaderInc(const char* header, bool system=false);
+std::string genMarshalRoutine(Type* t, bool out);
+std::string genMarshalPushRoutine(Type* t);
+std::string genMarshalPullRoutine(Type* t);
+std::string genServerDispatchSwitch(const std::vector<FnSymbol*>& fns);
+std::string genFuncToSetServerGlobals(void);
+std::string genDebugPrintCall(FnSymbol* fn, const char* pfx="");
+std::string genDebugPrintCall(const char* msg, const char* pfx="");
+std::string genFuncNumericID(FnSymbol* fn);
+std::string genServerWrapperCall(FnSymbol* fn);
+std::string genClientsideRPC(FnSymbol* fn);
+std::string genServersideRPC(FnSymbol* fn);
+std::string genMarshalCall(const char* s, const char* v, Type* t, bool out);
+std::string genMarshalPushCall(const char* s, const char* v, Type* t);
+std::string genMarshalPullCall(const char* s, const char* v, Type* t);
+std::string genTypeName(Type* t);
+std::string genSocketCall(const char* s, const char* v, const char* l,
+                          bool out);
+std::string genSocketCall(const char* s, const char* v, bool out);
+std::string genSocketPushCall(const char* s, const char* v);
+std::string genSocketPullCall(const char* s, const char* v);
+std::string genAddressOf(const char* var);
+std::string genAddressOf(std::string& var);
+std::string genSizeof(const char* var);
+std::string genSizeof(std::string& var);
+std::string genNewDecl(const char* t, const char* n);
+std::string genNewDecl(Type* t, const char* n);
+
+};
+
+//
+// This is the main entrypoint for MLI code generation, call this if the
+// necessary conditions are met in codegen().
+//
+void codegenMultiLocaleInteropWrappers(void) {
+  Vec<ModuleSymbol*> &mds = allModules;
+
+  // Insert all kinds of debug printlines into generated code for now.
+  MLIContext mli(true);
+
+  mli.emitClientPrelude();
+  mli.emitServerPrelude();
+
+  forv_Vec(ModuleSymbol, md, mds) {
+    mli.emit(md);
+  }
+
+  mli.emitMarshalRoutines();
+  mli.emitServerDispatchRoutine();
+
+  return;
+}
+
+MLIContext::MLIContext(bool injectDebugPrintlines, bool separateHeaders) {
+
+  // Yes, I know this isn't the most optimal way to initialize these!
+  this->injectDebugPrintlines = injectDebugPrintlines;
+  this->separateHeaders = separateHeaders;
+
+  openCFile(&this->fiMarshalling, gen_mli_marshalling, "c");
+  openCFile(&this->fiClientBundle, gen_mli_client_bundle, "c");
+  openCFile(&this->fiServerBundle, gen_mli_server_bundle, "c");
```

Comment:
I'm not sure. I don't know if the cost of opening the file just to get the path is that big of a deal, though?

Reply:
It's a system call (which is generally expensive because the operating system has to switch to a mode that allows that sort of operation, iirc) and it performs a slew of other operations.

According to [wikipedia](https://en.wikipedia.org/wiki/Open_(system_call)):

> During the open, the filesystem may allocate memory for buffers, or it may wait until the first operation.
> ...
> The absolute file path is resolved. This may include connecting to a remote host and notifying an operator that a removable medium is required. It may include the initialization of a communication device.

So if the user specified that they wanted the generated code to live across a mount point, for instance, this could be even more costly

> ...
> The first access to at least the directory within the filesystem is performed.
> ...
> If the file is expected to exist and it does, the file access, as restricted by permission flags within the file meta data or access control list, is validated against the requested type of operations. This usually requires an additional filesystem access although in some filesystems meta-flags may be part of the directory structure. 
> ...
> Various other errors which may occur during the open include directory update failures, un-permitted multiple connections, media failures, communication link failures and device failures. 
> ...
> In many cases programming language-specific run-time library opens may perform additional actions including initializing a run-time library structure related to the file.

This is way more work than just accessing a global variable that you had stored in your program

Reply:
It's true that it's a system call and that it's relatively more expensive than a userspace only operation. 

If I can find a way to get the same values without opening the file, I'll use that.